File: src/main/java/kr/mdns/madness/response/ApiResponse.java
package kr.mdns.madness.response;

import lombok.Getter;

@Getter
public class ApiResponse<T> {
    private final int code;
    private final String message;
    private final T data;

    // 생성자 숨기기 (외부에서 직접 인스턴스 생성 불가)
    public ApiResponse(int code, String message, T data) {
        this.code = code;
        this.message = message;
        this.data = data;
    }

    /**
     * 기본 성공 응답 생성
     * 
     * @param data 응답 데이터
     */
    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(0, "success", data);
    }

    /**
     * 커스텀 응답 생성
     * 
     * @param code    응답 코드
     * @param message 응답 메시지
     * @param data    응답 데이터
     */
    public static <T> ApiResponse<T> of(int code, String message, T data) {
        return new ApiResponse<>(code, message, data);
    }

    /**
     * 오류 응답 생성 (데이터 없음)
     * 
     * @param code    오류 코드
     * @param message 오류 메시지
     */
    public static <T> ApiResponse<T> failure(int code, String message) {
        return new ApiResponse<>(code, message, null);
    }
}

File: src/main/kotlin/api/madn/es/common/response/ApiResponse.kt
package api.madn.es.common.response

import api.madn.es.common.response.exception.ErrorCode
import com.fasterxml.jackson.annotation.JsonInclude
import java.time.OffsetDateTime

data class ApiResponse<T>(
    val success: Boolean,
    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    val data: T? = null,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    val error: ApiError? = null,

    val timestamp: OffsetDateTime = OffsetDateTime.now()
) {
    companion object {
        fun <T> success(data: T? = null): ApiResponse<T> =
            ApiResponse(success = true, data = data)

        fun failure(
            errorCode: ErrorCode,
            details: ErrorDetails? = null
        ): ApiResponse<Nothing> =
            ApiResponse(
                success = false,
                error = ApiError(
                    code = errorCode.code,
                    message = errorCode.message,
                    details = details
                )
            )
    }
}